crates/c-api/src/async.rs

```diff
@@ -0,0 +1,299 @@
+    c.config.async_stack_size(size);
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_context_epoch_deadline_async_yield_and_update(
+    mut store: CStoreContextMut<'_>,
+    delta: u64,
+) {
+    store.epoch_deadline_async_yield_and_update(delta);
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_context_out_of_fuel_async_yield(
+    mut store: CStoreContextMut<'_>,
+    injection_count: u64,
+    fuel_to_inject: u64,
+) {
+    store.out_of_fuel_async_yield(injection_count, fuel_to_inject);
+}
+
+pub type wasmtime_func_async_callback_t = extern "C" fn(
+    *mut c_void,
+    *mut wasmtime_caller_t,
+    *const wasmtime_val_t,
+    usize,
+    *mut wasmtime_val_t,
+    usize,
+    &mut Option<Box<wasm_trap_t>>,
+    *mut wasmtime_async_continuation_t,
+);
+
+#[repr(C)]
+pub struct wasmtime_async_continuation_t {
+    pub callback: wasmtime_func_async_continuation_callback_t,
+    pub env: *mut c_void,
+    pub finalizer: Option<extern "C" fn(*mut c_void)>,
+}
+
+unsafe impl Send for wasmtime_async_continuation_t {}
+unsafe impl Sync for wasmtime_async_continuation_t {}
+impl Drop for wasmtime_async_continuation_t {
+    fn drop(&mut self) {
+        if let Some(f) = self.finalizer {
+            f(self.env);
+        }
+    }
+}
+impl Future for wasmtime_async_continuation_t {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        let cb = this.callback;
+        if cb(this.env) {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+pub type wasmtime_func_async_continuation_callback_t = extern "C" fn(*mut c_void) -> bool;
+
+struct CallbackData {
+    env: *mut c_void,
+}
+unsafe impl Send for CallbackData {}
+
+async fn invoke_c_async_callback<'a>(
+    cb: wasmtime_func_async_callback_t,
+    data: CallbackData,
+    mut caller: Caller<'a, crate::StoreData>,
+    params: &'a [Val],
+    results: &'a mut [Val],
+) -> Result<()> {
+    // Convert `params/results` to `wasmtime_val_t`. Use the previous
+    // storage in `hostcall_val_storage` to help avoid allocations all the
+    // time.
+    let mut hostcall_val_storage = mem::take(&mut caller.data_mut().hostcall_val_storage);
+    debug_assert!(hostcall_val_storage.is_empty());
+    hostcall_val_storage.reserve(params.len() + results.len());
+    hostcall_val_storage.extend(params.iter().cloned().map(|p| wasmtime_val_t::from_val(p)));
+    hostcall_val_storage.extend((0..results.len()).map(|_| wasmtime_val_t {
+        kind: WASMTIME_I32,
+        of: wasmtime_val_union { i32: 0 },
+    }));
+    let (params, out_results) = hostcall_val_storage.split_at_mut(params.len());
+
+    // Invoke the C function pointer.
+    // The result will be a continutation which we will wrap in a Future.
+    let mut caller = wasmtime_caller_t { caller };
+    let mut trap = None;
+    let mut continuation = MaybeUninit::uninit();
```

Comment:
Due to this containing a function pointer I don't think that `uninit` is actually safe to do here, could this instead

Reply:
er, sorry, hit send to soon!

I take this back, I think this is fine as-is.

Reply:
Yeah I wanted to initialize a value, but you can't assign null to a function ptr in rust so this was the best I could come up with... 😄

Reply:
Yeah the risk of this is that it triggers UB if C accidentally doesn't fill in the structure, so the alternative would be to have a dummy function pointer which panics. I don't think it's worth it to go that far though, hence my reconsideration this is fine as is!

Reply:
I added the dummy function pointer - it's better to give guardrails.